### DISCUSSION TITLE
How to control device fail over to AP mode

### DISCUSSION THREAD

**Discussion by dkebler** [arendst/Tasmota#20152] [Q&A]
Here is my situation.  I live somewhere where the site power (from power company) now and then is lost.  (maybe 6-10 times a year more in the winter).

Thus I must pay attention to boot up robustness off all my network devices so that they all come back "online" without fail.  I have no general control over the order so devices must wait until the services they need come online.

One tasmota related issue per this situation is that some of the devices don't find the configured AP in time (because AP is also coming online)  and thus failover to AP mode (i.e. wifimanager) which makes then unreachable.  In this case I have to physically go to the device and restart it.    :(.

What I usually do is deploy with AP1 set to the AP it will connect to at deployment location.  Then I set AP2 to my lab AP so that if I pull the device and bring to lab it will still connect.   

**What I don't want is device to go into AP mode if it can't find either AP as this happens to some deployed devices after a site power loss for (APs are also restarting).**  

If this device won't connect fine I'll bring it to my lab.  If it still won't connect I'll reflash it. 

Going into failover AP mode might be ok but only if I can control a timeout length for that so to give my APs plenty of time to come online.

**So given all this what if any `WifiConfig` setting and/or other option (complied or otherwise) can I use to fit my desired configuration?**

-----
```
WifiConfig
0 = disable Wi-Fi Manager and reboot (used with alternate AP)
2 = set Wi-Fi Manager as the current configuration tool and start Wi-Fi Manager (web server at 192.168.4.1) for 3 minutes, then reboot and try to connect Wi-Fi network
4 = retry other AP without rebooting (default)
5 = wait until selected AP is available again without rebooting
6 = Wi-Fi parameters can only be entered via commands in the serial console
7 = set Wi-Fi Manager (web server at 192.168.4.1) as the current configuration tool restricted to reset settings only.
This setting is recommended for devices without an external control/reset button.
&emsp;No longer supported
1 = set SmartConfig (Android/iOS) for 3 minutes
3 = set WPS for 3 minutes
```

**Comment by sfromis** [comment]
I consider using `WifiConfig 2` bad config, unless you *specifically* want AP mode, and lack of security.

My experience is that if I power up a Tasmota device, and shortly after the AP, that Tasmota will connect within a short time. YMMV.

(For experimentation, you can have a Tasmota device act as an AP, using the range extender feature)

**Comment by jziolkowski** [comment]
I have default config, and I had tasmota devices reconnect to proper AP after many minutes of AP being offline. At worst, I needed to go to one, push the button and it somehow "forced" reconnection to wifi and mqtt.

**Reply by barbudor** [reply]
> What I usually do is deploy with AP1 set to the AP it will connect to at deployment location. Then I set AP2 to my lab AP so that if I pull the device and bring to lab it will still connect.

The default `WifiConfig 4` "retry other AP without rebooting (default)" seems to perfectly match your need
Tasmota will try alternatively AP1 and AP2 until one is up so it will cover both of your use cases (assuming the lab AP2 is not on by default).

If your mains doesn't come nicely but may come back and go multiple times, you should also probably disable ["Fast Power Cycle Device Recovery"](https://tasmota.github.io/docs/Device-Recovery/#fast-power-cycle-device-recovery) using `Setoption65 1`

**Comment by vic42** [comment]
`WifiConfig 4` and `SetOption65 1` would be my guess.

**Comment by dkebler** [comment]
My understanding  is that the default (`4`)  will: Try AP1, Try AP2,  keep trying for for some period (timeout) then set AP mode and reboot.

**Is that correct?**  If so in theory `4` works for me but only if I can then set that timeout value (_which currently I don't know what that is and AFAIK it's not settable_) as clearly it's not long enough.    

Hmmm....that makes me think.   Make a rule that just reboots the device if it can't connect after some time (say 20 secs)?   Anyway I just need something so device boot-up is more robust to mains loss.   

So, If I go with `2`  that might work assuming that after 3 minutes it will reboot using the currently configured APs and not clear them?   Still not the best to go into APmode from the get go but better than having to physically reset devices which is of course not possible if I am not around.

**Reply by barbudor** [reply]
4 never ever returns to AP
